services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusMigrationServiceTest.kt

```diff
@@ -0,0 +1,238 @@
+        }
+
+        @Test
+        fun `clears customers for an excavation notification`() {
+            val data =
+                ApplicationFactory.createExcavationNotificationData(
+                    customerWithContacts = ApplicationFactory.createCompanyCustomer().withContact(),
+                    contractorWithContacts =
+                        ApplicationFactory.createCompanyCustomer().withContact(),
+                    propertyDeveloperWithContacts =
+                        ApplicationFactory.createCompanyCustomer().withContact(),
+                    representativeWithContacts =
+                        ApplicationFactory.createCompanyCustomer().withContact(),
+                )
+
+            val response = HakemusMigrationService.clearCustomers(data)
+
+            assertThat(response.customersWithContacts()).isEmpty()
+        }
+    }
+
+    @Nested
+    @TestInstance(TestInstance.Lifecycle.PER_CLASS)
+    inner class ContactWithLeastMissingFields {
+        @Test
+        fun `returns null when the contact list is empty`() {
+            val response = listOf<Contact>().contactWithLeastMissingFields()
+
+            assertThat(response).isNull()
+        }
+
+        @Test
+        fun `returns the first element when the list has one element`() {
+            val contacts = listOf(ApplicationFactory.createContact())
+
+            val response = contacts.contactWithLeastMissingFields()
+
+            assertThat(response).isNotNull().isSameInstanceAs(contacts[0])
+        }
+
+        @Test
+        fun `returns the first element when the list has several elements with all fields`() {
+            val contacts =
+                listOf(
+                    ApplicationFactory.createContact(firstName = "Teppo"),
+                    ApplicationFactory.createContact(firstName = "Pekka"),
+                    ApplicationFactory.createContact(firstName = "Matti"),
+                    ApplicationFactory.createContact(firstName = "Hannu")
+                )
+
+            val response = contacts.contactWithLeastMissingFields()
+
+            assertThat(response).isNotNull().isSameInstanceAs(contacts[0])
+        }
+
+        @Test
+        fun `returns the element with the least missing values`() {
+            val contacts =
+                listOf(
+                    ApplicationFactory.createContact(firstName = null, lastName = "", phone = " "),
+                    ApplicationFactory.createContact(firstName = "Pekka", lastName = " "),
+                    ApplicationFactory.createContact(firstName = "Matti", email = "", phone = null),
+                    ApplicationFactory.createContact(
+                        firstName = "Hannu",
+                        lastName = null,
+                        email = " "
+                    )
+                )
+
+            val response = contacts.contactWithLeastMissingFields()
+
+            assertThat(response).isNotNull().isSameInstanceAs(contacts[1])
```

Comment:
So, it is Pekka? Reading just this test doesn't make it very clear how we have came into that conclusion. The names have some priority over phone? None of the contacts have email address, should there be such a test object too?

Reply:
Well, nevermid, this is not so important and you can ignore this. I thought that starting reading from the tests would make the logic more clear but it turned out to be the other way round.

Reply:
I changed the test to create a `Contact` directly, without using the factory method. I think this makes it easier to see which fields are missing and which are not. I also made the contacts named values, to give them some context.

Reply:
Thank you!